sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/SupportsPushDownAggregates.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.read;
+
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.sources.Aggregation;
+
+/**
+ * A mix-in interface for {@link ScanBuilder}. Data sources can implement this interface to
+ * push down aggregates to the data source.
+ *
+ * @since 3.2.0
+ */
+@Evolving
+public interface SupportsPushDownAggregates extends ScanBuilder {
```

Comment:
Some thoughts about the v2 pushdown API:

I've been reluctant to add more v2 pushdown APIs because the semantic is not clear. For now, we only have filter pushdown and column pruning, which is fine as these two don't conflict, the order doesn't matter. However, if we add more pushdown API, like limit pushdown, then the order matters. e.g. LIMIT + FILTER is different from FILTER + LIMIT.

I've been thinking about it many times. We can require the pushdown implementation to be stateful, and remember the order of the API calls. However, I found this makes the implementation very complicated, and it's simpler to just write catalyst rules.

Another idea is to only push down one operator, and apply column pruning at the end. e.g. if a data source supports both filter and limit pushdown, only one can be pushed down each time (the one closer to the scan node gets pushed). This way the implementation can still be simple, and people can write catalyst rules if they want to support complicated cases.

In general, I find it's more natural to implement SQL operators pushdown using tree traversal. The DS v2 pushdown API gets complicated very quickly when we add more features. One option is to have a Java version of `LogicalPlan` and provide convenient tree traversal APIs. But I'm not sure how many users will buy in instead of just writing catalyst rules.

Reply:
For this PR, I don't think a v2 aggregate pushdown API makes sense. File source v1 is still the default and the effort of this PR is wasted if we can't turn on file source v2 by default.

Can we simply write a catalyst rule to match Aggregate and file scan node, and implement the aggregate pushdown?

Reply:
@cloud-fan 
I quickly tried it out by adding a rule to match Aggregate and V2Relation. I did the following:
1. deleted the `SupportsPushDownAggregates` interface
2. added `pushAggregation` in `Scan`
```
  default void pushAggregation(Aggregation aggregation) {
    throw new UnsupportedOperationException(description() +
        ": Push down Aggregation is not supported");
  }
```
3. implemented the above method in `ParquetScan`
4. added a rule `PartialAggregatePushDown` in core/sql. I didn't add this in catalyst because I need to access some of the files in core/sql. 
5. In `PartialAggregatePushDown`, I only enable the push down if the table in V2Relation is a `ParquetTable`, because I only implemented `pushAggregation` in `ParquetScan` now.

Could you please take a look to see if this is what you want? Thank you very much!

Reply:
Can we support file source v1 first? That's the default.

Reply:
@cloud-fan 
Most of the customers at my work use DS v2. Is it OK with you that we use this PR to implement v2 support first, and I will have a follow up to add v1 support later? Thanks a lot!